ragna/assistants/_openai_api_compatible.py

```diff
@@ -0,0 +1,74 @@
+import json
+import os
+from typing import AsyncIterator, cast
+
+import httpx
+
+import ragna
+from ragna.core import PackageRequirement, Requirement, Source
+
+from ._api import ApiAssistant
+
+
+class OpenAIApiCompatible(ApiAssistant):
+    def __init__(self) -> None:
+        self._client = httpx.AsyncClient(
+            headers={"User-Agent": f"{ragna.__version__}/{self}"},
+            timeout=60,
+        )
+        self._base_url = os.environ.get("BASE_URL")
+
+    @classmethod
+    def requirements(cls) -> list[Requirement]:
+        return [PackageRequirement("httpx_sse")]
+
+    @classmethod
+    def display_name(cls) -> str:
+        return "OpenAIApiCompatible"
+
+    def _make_system_content(self, sources: list[Source]) -> str:
+        instruction = (
+            "You are an helpful assistants that answers user questions given the context below. "
+            "If you don't know the answer, just say so. Don't try to make up an answer. "
+            "Only use the sources below to generate the answer."
+        )
+        return instruction + "\n\n".join(source.content for source in sources)
+
+    async def _call_api(
+        self, prompt: str, sources: list[Source], *, max_new_tokens: int
+    ) -> AsyncIterator[str]:
+        import httpx_sse
+
+        async with httpx_sse.aconnect_sse(
+            self._client,
+            "POST",
+            f"{self._base_url}/v1/chat/completions",
```

Comment:
Azure OpenAI needs a [different scheme](https://learn.microsoft.com/en-us/azure/ai-services/openai/chatgpt-quickstart?tabs=command-line%2Cpython-new&pivots=rest-api#rest-api):

```
$AZURE_OPENAI_ENDPOINT/openai/deployments/gpt-35-turbo/chat/completions?api-version=2024-02-01
```

Reply:
It is compliant. You just need to use a different URL as posted above. The `gpt-35-turbo` in there is the name of the deployment, e.g. the model name. Meaning, similar to the llamafile approach in this PR, you don't pass a model in the request (see https://github.com/Quansight/ragna/pull/424#discussion_r1615584286). Apart from that, this is exactly the same API as the public OpenAI one.

Reply:
> It is compliant.

My understanding was that it is not compliant. Can you point me to some resources on this please? 

As a quick check, I went to the OpenAPI spec for Azure, and I saw [this](https://github.com/Azure/azure-rest-api-specs/blob/c3cc9abe085093ba880ee3eeb792edb4fa789553/specification/cognitiveservices/data-plane/AzureOpenAI/inference/stable/2024-02-01/inference.json#L1225) example for the `/chat/completions` endpoint, which appears to not be compliant. A quick skim of the Azure API documentation also seems to suggest [several](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference) other examples.

Reply:
> My understanding was that it is not compliant. Can you point me to some resources on this please?

I can't. We have used it in an environment that I cannot share publicly.

> As a quick check, I went to the OpenAPI spec for Azure, and I saw [this](https://github.com/Azure/azure-rest-api-specs/blob/c3cc9abe085093ba880ee3eeb792edb4fa789553/specification/cognitiveservices/data-plane/AzureOpenAI/inference/stable/2024-02-01/inference.json#L1225) example for the `/chat/completions` endpoint, which appears to not be compliant.

IIUC, these are just extra parameters that Azure OpenAI supports but OpenAI does not. Meaning Azure OpenAI seems to be a superset of the public API. I can attest to the fact that you do not need to set any of these parameters to use it.

> A quick skim of the Azure API documentation also seems to suggest [several](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference) other examples.

I did not see anything that would make it non-compliant. Could you share the specific section / example that you are referring to?

Reply:
> Meaning Azure OpenAI seems to be a superset of the public API.
Maybe this reduces to me having a different (incorrect?) definition of what a compliant API is then. I would have assumed that a superset automatically made it non-compliant.

> I did not see anything that would make it non-compliant.
They had stuff related to an `operations` [endpoint](https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#get-a-generated-image-result-dall-e-2-preview) that I didn't recognise on a quick glance. Having a version number as a query parameter also seemed like it could create compatibility problems, unless things were deprecated/added at the same time for both API endpoints, which they aren't. But again, maybe I have a different or incorrect understanding of what compatibility means here.